### ISSUE TITLE
The adjacency matrix of GAT model

### ISSUE THREAD

**Issue by eeluo** [dmlc/dgl#2877]
Does the GAT uses adjacency matrix?I saw other code have an adjacency matrix to input, but I haven't seen that yet in the dgl.
And also,can I uses adjacency matrix which calculated by myself to the input of Gat model?

**Comment by jermainewang** [comment]
You can safely view `DGLGraph` as the equivalence of the adjacency matrix. Therefore, a function on adj matrix becomes a function on `DGLGraph` in DGL. Let's now look at the code of `GATConv`:
https://github.com/dmlc/dgl/blob/2a3a9a00117ec022d801505b858d21abea3f10dc/python/dgl/nn/pytorch/conv/gatconv.py#L283-L305

You could see on L298 there is a call to `graph.apply_edges` and on L301 there is a call to `edge_softmax(graph, e)`, both having `graph` as the function inputs.

To use your own adj matrix, you should create a graph from it and feed it to `GATConv`. If you want to further incorporate the edge weight into the model, there are multiple options. One would be after L299, scale the unnormalized attention values `e` with your edge weight before softmax.

**Comment by jermainewang** [comment]
> So as you say, if I want to change my own adj matrix to calculate, I should first create a graph form it and feed the graph to "GATConv". Therefore, if I want to change the adj matrix (same graph but different adj matrix) in every per iteration, I should recreate the graph every time, Is that so?

Yes

> And if I want to create a graph from adj matrix, I can only use the functions such as "g.add_nodes(num) g.add_edges(src, dst)" every time when I want to recerate a graph?

There are a bunch of operators for creating a graph. https://docs.dgl.ai/api/python/dgl.html#graph-create-ops

**Comment by eeluo** [comment]
> 
> 
> > So as you say, if I want to change my own adj matrix to calculate, I should first create a graph form it and feed the graph to "GATConv". Therefore, if I want to change the adj matrix (same graph but different adj matrix) in every per iteration, I should recreate the graph every time, Is that so?
> 
> Yes
> 
> > And if I want to create a graph from adj matrix, I can only use the functions such as "g.add_nodes(num) g.add_edges(src, dst)" every time when I want to recerate a graph?
> 
> There are a bunch of operators for creating a graph. https://docs.dgl.ai/api/python/dgl.html#graph-create-ops


I got it，thank you so much!
